### PR TITLE
Create, delete and update songs

### PR DESCRIPTION
**Pull request by alfredgrip**
This PR adds functionality to create, delete and update songs from the db.

A page for each song is added. This is currently done with the title as the parameter. A problem however is that title is not marked as unique in db (maybe it should be, maybe it shouldn't). Each song has an ID which is unique, but to make it look neat in the URL (and for SEO maybe?) I chose to use the title instead.

Currently, infu.mastare and infu.arkivarie can create, update and delete songs. This ofc can be discussed and adjusted.

When an authenticated user with correct access policieswants to delete a song, currently they get a bunch of warnings to confirm their action. This is done by basic confirmation, maybe a modal could be shown instead, but then again this functionality will only be seen by a few users. Another discussion point, should we be able to delete songs that easily? The archive should be treated with care and IMO we should never delete a song unless it was uploaded by mistake. Should we instead have to manually query the db to delete songs and remove the ability to delete songs?

For reviewer: please double check any logic and practices in the backend, this is not my strong side.

Known issue: ugly rendering occurs for some characters, don't really know how to fix this.
Screenshot 2024-01-12 at 21 49 43

Future work: add ability for everyone to request a song to be created (like news articles)?

### CODE REVIEW

**Comment by danieladugyan** [comment]
Great addition! 🤩 
I'll look over the details as soon as possible, but I can respond immediately below to your observations.

**Comment by danieladugyan** [comment]
> A page for each song is added. This is currently done with the title as the parameter. A problem however is that title is not marked as unique in db (maybe it should be, maybe it shouldn't). Each song has an ID which is unique, but to make it look neat in the URL (and for SEO maybe?) I chose to use the title instead.

We have the same problem for news articles -- there's no guarantee that the title is unique and the ID is kind of ugly. We solve it by generating a slug for each article based on the title. For instance, `TIFs Squashstege` becomes `tifs-squashstege-1` (_sidenote: the `1` seems unnecessary?_). I think that solution works equally well here, but obviously it's a bit more effort.

**Comment by danieladugyan** [comment]
> Another discussion point, should we be able to delete songs that easily? The archive should be treated with care and IMO we should never delete a song unless it was uploaded by mistake. Should we instead have to manually query the db to delete songs and remove the ability to delete songs?

This is an important point. We should never actually delete songs from the database. Again, we can draw inspiration from how news articles are handled. Instead of actually deleting them when they're deleted in the GUI, we perform a _soft deletion_, which simply means updating the row in question to have a `deleted: true` field. In practice, we usually use `deletedAt: date` instead, but the principle remains the same. It's implemented as follows:

```ts
prisma.song.update({
   where: { id: data.id },
   data: { deletedAt: new Date() },
});
```
When fetching songs we make sure we only include songs that haven't been deleted:
```ts
prisma.song.findMany({ where: { deletedAt: null } }}
```

**Comment by danieladugyan** [comment]
> Known issue: ugly rendering occurs for some characters, don't really know how to fix this.

I think this should be solved by using `@html`. Not sure how these HTML characters interact with the search logic though, I suspect, for instance, that searching for `Don't think twice` won't find the song seen in your screenshot. Not sure though. 
_On another sidenote, using all these special `@html` tags makes me a bit uneasy, we're opening ourselves up for XSS attacks if we forget to sanitize the input._

**Comment by danieladugyan** [comment]
Lots of observations and lots of answers 😅, obviously feel free to do however much/little you feel like -- I can pick-up whatever's left before merging.

**Comment by alfredgrip** [comment]
Thanks for the great input! Seems very reasonable to generate a slug here as well. As for deletion, that's a good point and much more reasonable than deleting it from the db. As for the ugly rendering, I'll try a bit more and see if I come up with a nice and safe solution. I'll work on this a bit more!

**Comment by alfredgrip** [comment]
After looking into it a bit more, I guess this needs a migration, maybe you can take a look at that 😅

**Comment by danieladugyan** [comment]
If I never have to do a migration like that again it'll be too soon -- that was awful. Anyways, there's a migration up now.

**Comment by alfredgrip** [comment]
A big thanks for fixing the migration! ❤️ I've done some changes now to account for using the slug instead, fixing deletion och songs and adding ability to restore a song!

Side note about the ugly rendering for some melodies: I noticed that I could edit the song's melody, and manually replace `&#039;` with a `'`. After that it rendered fine so I guess this isn't a problem if we just edit them.

**Review by danieladugyan — APPROVED** [review]
I refactored some details -- left comments above and in commit messages. As long as you don't have any objections I think we can merge!

File: src/routes/[[lang=lang]]/songbook/+page.svelte [hunk 1]
```diff
@@ -21,9 +26,43 @@
...
+      on:click={async () => {
+        const urlParams = new URLSearchParams(data.params);
+        if (urlParams.has("show-deleted")) {
+          urlParams.delete("show-deleted");
+        } else {
+          urlParams.set("show-deleted", "true");
+        }
+        urlParams.delete("page"); // Reset the page counter
+        await goto(`?${urlParams.toString()}`, { replaceState: true });
+      }}
```
[the remark this answers is not in the document]
Comment (alfredgrip) [reply src/routes/[[lang=lang]]/songbook/+page.svelte]: Although your new way to do this is neat and simple, this does a page refresh each time we apply a new filter (including categories) which also resets our scroll position. These changes also forces us to manually press enter after we input a search to request it, which is quite annoying. Is there a way to combine these to get best of both worlds?

**Comment by alfredgrip** [comment]
Nice refactoring! I added some cancellation buttons but other than that I think it's basically good to go. However I did leave a comment on one of your reviews, so please check that out. Also I noticed that when a user saves an edit on a song, this gets printed to the server `Unknown error occured Redirect { status: 303, location: '/songbook/min-nya-l%C3%A5t-2' }` (or another location depending on which song was edited). Can't really figure out why and it would be nice to resolve.

File: src/routes/[[lang=lang]]/songbook/+page.svelte (hunk 1 above)
[the remark this answers is not in the document]
Comment (alfredgrip) [reply src/routes/[[lang=lang]]/songbook/+page.svelte]: Also, if you first set a category, then enter som search input and press enter, the category gets reset which is unpleasant. This doesn't occur if you do it in the other order, i.e. first search, hit enter and then set a category.

[the remark this answers is not in the document]
Comment (danieladugyan) [reply src/routes/[[lang=lang]]/songbook/+page.svelte]: Apparently I'm just bad at SvelteKit forms 😅 should work properly now.

**Comment by danieladugyan** [comment]
> Also I noticed that when a user saves an edit on a song, this gets printed to the server `Unknown error occured Redirect { status: 303, location: '/songbook/min-nya-l%C3%A5t-2' }` (or another location depending on which song was edited). Can't really figure out why and it would be nice to resolve.

"Fixed" by altering the logic that prints out this error message. To be honest I'm not quite sure what that code is supposed to achieve.

**Comment by alfredgrip** [comment]
Great changes! I'm happy as it is now and think it's ready to merge! Just one thing: should we populate the `updatedAt` field when we create a song or not? I noticed that this isn't how news are handled, i.e. updatedAt is null when we create something for the first time.

**Comment by danieladugyan** [comment]
> Just one thing: should we populate the `updatedAt` field when we create a song or not? I noticed that this isn't how news are handled, i.e. updatedAt is null when we create something for the first time.

Leaving it empty seems fine to me, although to be honest I don't really have a strong preference. Populating it doesn't seem to really do much? If it's empty by default it at least serves as a `hasBeenEdited` field.

**Pull request merged**